### DISCUSSION TITLE
Volar 2 - Two servers, only one for completion

### DISCUSSION THREAD

**Discussion by DanielViberg** [yegappan/lsp#471] [Q&A] [answered]
Hello,

I'm far out in deep waters in my assumptions so please correct me if im wrong. The new volar 2 server depends on two lsp's running side by side. One for vue completion <style> and "\<template\>" and tsserver for <script>. 

It seems like some have managed to make it work i neovim, i've tried to update my config for this client: 
```json
{
      "name": "ts-ls",
      "filetype": ["typescript", "javascript", "vue"],
      "path": "typescript-language-server",
      "args": ["--stdio"],
      "features": {
        "documentFormatting": false
      },
      "initializationOptions": {
          "plugins":[
            {
              "name": "@vue/typescript-plugin",
              "location":"home/daniel/.nvm/versions/node/v20.11.0/lib/node_modules/@vue/typescript-plugin",
              "languages": ["vue"]
            }
          ]  
      }
    },
    {
      "name": "vue-ls",
      "filetype": ["vue"],
      "path": "vue-language-server",
      "args": ["--stdio"],
      "features": {
        "documentFormatting": true
      },
      "initializationOptions": {
          "typescript": {
              "tsdk": "/home/daniel/.nvm/versions/node/v20.11.0/lib/node_modules/typescript/lib"
          }
      },
      "workspaceConfig": {
        "html": {
          "format": {
            "wrapAttributes": "force-aligned"
          }
        }
      }
    },

```
The problem i have is that i can either get only vue (html) completions or ts completions. But not both. I read section 16 in the doc, though did not find how to use the same feature from two server. (Maybe i shouldnt?)
 **With the current config I'm getting ts completions, if i put the vue server first i get vue completions, how do i get both?**
Could this be a client feature limitation, or is it maybe an issue with my config?

I should note that there is an discussion ongoing about the correct way of configuring volar:
https://github.com/vuejs/language-tools/issues/3925#issuecomment-1988737023

**Comment by andlrc** [comment]
Hey Daniel

It's been a while since I made the initial PR's that added support for multiple servers. And I havn't followed to closely lately, but when I left it off only supported completions from one LSP-server. I think it's written quite clear in the docs:

> By default the language server defined first will be used for as much as it
supports, then the next and so on. With the exception that diagnostics from
all running language servers will be combined.

I guess we could be more clear about the fact that everything but diagnostics are being used from one LSP-server only.

I also think that it would make sense for completions to come from multiple sources.

Looking at the source code it stills looks like only one server is used for completion, see

https://github.com/yegappan/lsp/blob/0b9bba0ee4eaa84a75d88a0775b3b34dbf42b12d/autoload/lsp/completion.vim#L471
and
https://github.com/yegappan/lsp/blob/0b9bba0ee4eaa84a75d88a0775b3b34dbf42b12d/autoload/lsp/completion.vim#L539

and 
https://github.com/yegappan/lsp/blob/0b9bba0ee4eaa84a75d88a0775b3b34dbf42b12d/autoload/lsp/completion.vim#L565
for 24/7 completion 


/ Andreas

**Reply by andlrc** [reply]
Oh this line is also related to completion:

https://github.com/yegappan/lsp/blob/0b9bba0ee4eaa84a75d88a0775b3b34dbf42b12d/autoload/lsp/lsp.vim#L415

**Comment by DanielViberg** [comment]
I've managed to get completion of multiple servers working, so far it seems stable. https://github.com/DanielViberg/lsp
I'll continue verifying the stability, if possible please let me know if you see any issues in my changes.

**Comment by DanielViberg** [comment]
Currently working on:
```
Error: request workspace/executeCommand failed (Unhandled method workspace/executeCommand, error = MethodNotFound)
Error: request completionItem/resolve failed (Request completionItem/resolve failed with message: Cannot read properties of undefined (reading '1'), error = InternalError)
```

**Comment by DanielViberg** [comment, marked as answer]
https://github.com/yegappan/lsp/pull/480
